Backward compatibility with Titan

rafatuita opened the pull request:
Implemented backward compatibility with titan and its previous versions compatible with JanusGraph's storage format.

Addresses issue #73.

mbrukman commented:
@rafatuita – sorry about that, @janusgraph-bot wasn't yet aware of your CLA signature; that's now fixed.

mbrukman commented on CONTRIBUTORS.txt:
```diff
@@ -78,6 +78,7 @@ Peter Beaman <pbeaman@akiban.com>
 Petter Aas <aaspet@gmail.com> <petter.aas@fronter.com>
 Pieter <pieter@pieter-HP-EliteBook-8570w.(none)>
 PommeVerte <dylan.millikin@gmail.com>
+Rafael Fernandes <luizrafael@gmail.com>
```
You should add yourself to `AUTHORS.txt` as well since you retain copyright over your contributions.

rafatuita replied:
ok, done.

mbrukman commented on janusgraph-cassandra/src/test/java/org/janusgraph/graphdb/CassandraGraphTest.java:
```diff
@@ -106,4 +113,29 @@ public void testCustomConfigUsedByTx() {
+                ));
+
+        graph = (StandardJanusGraph) JanusGraphFactory.open(wc);
+
```
Extra blank line.

rafatuita replied:
done.

mbrukman commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/JanusGraphConstants.java:
```diff
@@ -42,6 +42,11 @@
+    /**
+     * Past versions of Titan Graph with which the runtime version shares a compatible storage format
+     */
+    public static final List<String> TITAN_BACKWARD_COMPATIBLE_VERSIONS;
```
For conciseness and readability, I would just drop the `BACKWARD_` part — yes, we're only compatible with versions of Titan in the past, but it would read nicer (I think) if we just have `TITAN_COMPATIBLE_VERSIONS` and `JANUSGRAPH_COMPATIBLE_VERSIONS` and similarly `titan.compatible-versions` and `janusgraph.compatible-versions`.

rafatuita replied:
done

mbrukman commented on pom.xml:
```diff
@@ -64,6 +64,7 @@
     </scm>
     <properties>
         <janusgraph.compatible.versions />
```
Doesn't this field have to list the current version, i.e., `0.1.0-SNAPSHOT`, to be compatible with JanusGraph itself, at the same version? Otherwise, if you run JanusGraph the first time to create a graph, then stop JanusGraph, and start it up again, it will not be compatible with itself.

It doesn't appear that self-version is automatically compatible with itself from looking at the code, unless I'm missing it.

Have you tested this change? Does it fix the compatibility issues with Titan, and does it work with JanusGraph-created graphs as well?

rafatuita replied:
hi @mbrukman, I wondered the same thing, but the following code checks first thing current version..

```java
if (!JanusGraphConstants.VERSION.equals(version) && !JanusGraphConstants.COMPATIBLE_VERSIONS.contains(version)) {
```

but just in case, I brought up a brand new empty cassandra install and ran quite a few tests, works just fine.

pluradj replied:
Current JanusGraph version is set on graph initialization here https://github.com/rafatuita/janusgraph/blob/3101295393816df4c178678c2b29771954f6e3b3/janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java#L1367

mbrukman commented:
@rafatuita – please also squash your commits into a single commit.

rafatuita commented:
@mbrukman about squashing, not really savvy enough to do this, how do I do it? Thought squashing would happen accepting the PR, I might be using this wrong for a while :)

jerryjch commented:
>> Thought squashing would happen accepting the PR,
Yes, it can be done when merging the PR.

rafatuita commented:
@mbrukman just realized I pushed a fix for issue #60 on this PR by mistake, was supposed to be a new PR, what can we do in this case? by the way, fixed the code by your last comment

mbrukman commented:
@rafatuita – you can do:

```bash
% git checkout -b new-temp-branch # to be used for another PR
% git log  # find the hash of the most recent commit, you'll need it later

% git checkout old-branch  # used for this PR
% git reset --hard HEAD~1   # walk back one commit
% git push -f    # update PR
```

Then:

```bash
% git checkout master
% git checkout -b new-branch-for-pr   # choose a better name :-)
% git cherry-pick [hash-of-commit-from-above]
% git push   # send this for review, create a new PR
```

That should do it. There may be other / more simpler ways of doing this, but that's how I would proceed.

amcp commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/JanusGraphConstants.java:
```diff
@@ -73,6 +78,11 @@
         }
         COMPATIBLE_VERSIONS = b.build();
+        
+        for (String v : props.getProperty("titan.compatible-versions", "").split(",")) {
```
stream().map().filter().collect(Collectors.toList())

rafatuita replied:
Sorry but didn't quite get that, I actually followed the previous implementation which seems to work just fine, can you please elaborate bit better? thanks.

amcp replied:
b.addAll(props.getProperty("titan.compatible-versions", "").split(",").stream().map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toList()));

rafatuita replied:
I got that sure. I just followed previous implementation, shall I change from line 74 til 86 with this? Titan had this before so I just followed it to keep it consistent.

amcp replied:
I guess I like to pick up these changes as we go, especially seeing as the right side of the colon in the original for loop does a lot of work to get to something iterable. Your choice, im approving.

rafatuita replied:
agreed! I'll make the proper changes (there's a conflict as well), submit and squash commits.

amcp commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java:
```diff
@@ -1381,10 +1385,14 @@ public GraphDatabaseConfiguration(ReadConfiguration localConfig) {
+                    String version = globalWrite.get(INITIAL_JANUSGRAPH_VERSION);
+                    Preconditions.checkArgument(version!=null,"JanusGraph version has not been initialized");
+                    if (!JanusGraphConstants.VERSION.equals(version) && !JanusGraphConstants.COMPATIBLE_VERSIONS.contains(version)) {
+                        throw new JanusGraphException("StorageBackend version is incompatible with current JanusGraph version: storage=" + version + " vs. runtime=" + JanusGraphConstants.VERSION);
```
lets make a string constant and reuse the error message

amcp commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java:
```diff
@@ -1381,10 +1385,14 @@ public GraphDatabaseConfiguration(ReadConfiguration localConfig) {
-                    throw new JanusGraphException("StorageBackend version is incompatible with current JanusGraph version: storage=" + version + " vs. runtime=" + JanusGraphConstants.VERSION);
+                try {
+                    String version = globalWrite.get(INITIAL_JANUSGRAPH_VERSION);
+                    Preconditions.checkArgument(version!=null,"JanusGraph version has not been initialized");
```
share this message with the version != null check below?

rafatuita replied:
this is how it looks now...

```java
Exception in thread "main" org.janusgraph.core.JanusGraphException: StorageBackend version is incompatible with current JanusGraph version: storage [0.0.0] vs. runtime [0.1.0-SNAPSHOT]
	at org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration.checkBackwardCompatibilityWithTitan(GraphDatabaseConfiguration.java:1490)
	at org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration.<init>(GraphDatabaseConfiguration.java:1395)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:108)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:98)
	at org.janusgraph.core.JanusGraphFactory$Builder.open(JanusGraphFactory.java:153)
```

amcp replied:
thanks!

amcp commented:
lgtm, please rebase and then good to go

amcp commented:
@rafatuita to squash, `git rebase -i HEAD~3` and replace "pick" on the second and third lines with "s" (for squash). Then `git push -f` to update your branch.

mbrukman commented:
@rafatuita — not sure what happened with the CLA there, but rather than using `git merge`, could you please use `git rebase master` on your branch instead, so it only contains your commits and nothing else? See my instructions above.

pluradj commented:
@mbrukman I screwed up the CLA check -- I merged a conflict in the root `pom.xml` using the GitHub UI, and it didn't use my CCLA email address. Not sure how to revert my commit.

mbrukman commented:
@pluradj – thanks for clarification.

@rafatuita – you can fix this via the `git` CLI, please?

rafatuita commented:
@mbrukman looks good now I'd say... I've reword'ed the commit message as well.

mbrukman commented:
@rafatuita – the GitHub view says: "This branch has conflicts that must be resolved."

Please run:

```bash
$ git checkout master
$ git pull upstream master
$ git checkout configuration
$ git rebase master
```

At this point, you'll find that you have conflicts. Fix the conflict in `pom.xml`, save the file, then run:

```bash
$ git add pom.xml
$ git rebase --continue
```

And you should be done. Then, update this PR:

```bash
$ git push -f
```

rafatuita commented:
@mbrukman I don't see any conflict on pom.xml, I believe @pluradj had fixed that

```
admins-MacBook-Pro:janusgraph admin$ git checkout master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
```
```
admins-MacBook-Pro:janusgraph admin$ git pull upstream master
Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```
```
admins-MacBook-Pro:janusgraph admin$ git checkout master
Already on 'master'
Your branch is up-to-date with 'origin/master'.
```
```
admins-MacBook-Pro:janusgraph admin$ git pull upstream master
Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```
```
admins-MacBook-Pro:janusgraph admin$ git checkout configuration
Switched to branch 'configuration'
Your branch is up-to-date with 'origin/configuration'.
```
```
admins-MacBook-Pro:janusgraph admin$ git rebase master
Current branch configuration is up to date.
```

don't see anything here, from here onwards, nothing happens.
```
admins-MacBook-Pro:janusgraph admin$ vim pom.xml 
```
```
admins-MacBook-Pro:janusgraph admin$ git add pom.xml
```
```
admins-MacBook-Pro:janusgraph admin$ git rebase --continue
No rebase in progress?
```
```
admins-MacBook-Pro:janusgraph admin$ git push -f
Everything up-to-date
```

mbrukman commented:
@rafatuita – this is the problem:

> ```bash
> admins-MacBook-Pro:janusgraph admin$ git pull upstream master
> Permission denied (publickey).
> fatal: Could not read from remote repository.
> 
> Please make sure you have the correct access rights
> and the repository exists.
> ```

As this error shows, you didn't actually sync your local repo because your SSH key was not available, thus the rest of the commands have no output.

You probably didn't run `ssh-add` to add your GitHub SSH key, or you're not even running the SSH agent (or didn't set the env vars) in this shell session.

sjudeng commented:
@mbrukman Since it's so minor maybe in this case you can just resolve the conflict as part of your merge commit?

pluradj commented:
@sjudeng Off-topic: I sent you a couple emails over the past week -- did you see them? Would you be able to send a reply via email? Thanks.

rafatuita commented:
@mbrukman done.

mbrukman commented:
@rafatuita -- thanks for squashing and re-pushing the change! I'll wait for the Travis run on your new PR version to complete and merge once it's green.
